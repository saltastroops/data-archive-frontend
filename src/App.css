.dimScreen {
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  height: 100%;
  width: 100%;
  margin: 0;
  padding: 0;
  background: #000000;
  opacity: 0.5;
  filter: alpha(opacity=30);
  -moz-opacity: 0.3;
  z-index: 101;
}
.signal {
  border: 15px solid #fff;
  border-radius: 30px;
  height: 130px;
  left: 50%;
  margin: -15px 0 0 -15px;
  opacity: 0;
  position: absolute;
  top: 50%;
  width: 130px;

  animation: pulsate 1s ease-out;
  animation-iteration-count: infinite;
}

@keyframes pulsate {
  0% {
    transform: scale(0.1);
    opacity: 0;
  }
  50% {
    opacity: 2;
  }
  100% {
    transform: scale(3);
    opacity: 0;
  }
}

.search-results.table {
  margin-bottom: 20px;
}

.search-results.cell {
  display: flex;
  align-items: center;
  padding-left: 5px;
  white-space: nowrap;
}

.search-results.cell.header {
  background-color: white;
  border-bottom: 1px solid lightgray;
  cursor: pointer;
}

.search-results.cell.header.cart {
  cursor: text;
}

.search-results.cell span {
  overflow: hidden;
  text-overflow: ellipsis;
}

.search-results.observation-header {
  background-color: #f2fffb;
  border-bottom: 1px solid lightgray;
  border-top: 2px solid lightgray;
}

.search-results.file-row-even {
  background-color: rgb(245, 245, 245);
}

.search-results.file-row-odd {
  background-color: white;
}

.ReactVirtualized__Grid {
  outline: none;
}
