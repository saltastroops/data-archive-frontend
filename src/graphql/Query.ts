import gql from "graphql-tag";

// TODO UPDATE
/**
 * Retrieving the user information.
 */
const USER_QUERY = gql`
  query USER_QUERY {
    user {
      familyName
      givenName
      roles
    }
  }
`;

export { USER_QUERY };
