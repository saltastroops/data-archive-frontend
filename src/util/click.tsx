import { ReactWrapper } from "enzyme";
import * as React from "react";

/**
 * Call the onClick method of the element wrapped in a wrapper with a mouse
 * click event.
 *
 * This function is intended for use in unit tests only. It is required because
 * the standard  Enzyme simulate does not seem to work for routing (see
 * https://github.com/ReactTraining/react-router/issues/4337).
 *
 * Properties
 * ----------
 * wrapper:
 *     Enzyme wrapper, which should wrap exactly one element.
 */
export default function click(
  wrapper: ReactWrapper<any, React.Component, any>
) {
  wrapper.props().onClick(new MouseEvent("click"));
}
