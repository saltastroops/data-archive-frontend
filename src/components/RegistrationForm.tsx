import { validate } from "isemail";
import * as _ from "lodash";
import * as React from "react";
import { Mutation } from "react-apollo";
import { Redirect } from "react-router";
import styled from "styled-components";
import { SIGNUP_MUTATION } from "../graphql/Mutations";
import InputField from "./basicComponents/InputField";

/**
 * Input for the registration form.
 *
 * Properties:
 * -----------
 * affiliation:
 *     The affiliation of the user, such as a university or an institute.
 * confirmPassword:
 *      A matching password that must match with the original password.
 * email:
 *     Email address. This will be stored in lower case.
 * familyName:
 *     The family name (surname).
 * givenName:
 *     The given name (first name).
 * password:
 *     The password, which must have at least 7 characters.
 * username:
 *     The username, which must not contain upper case letters.
 */
export interface IRegistrationFormInput {
  affiliation: string;
  confirmPassword: string;
  email: string;
  givenName: string;
  familyName: string;
  password: string;
  username: string;
}

/**
 * The state of the registration form.
 *
 * Properties:
 * -----------
 * errors:
 *     An object of keys and error messages.
 * registered:
 *     Whether the registration was successful.
 * userInput:
 *     Values input by the user.
 */
interface IRegistrationFormState {
  errors: Partial<IRegistrationFormInput> & { responseError?: string };
  registered: boolean;
  userInput: IRegistrationFormInput;
}

/**
 * The cache for the registration form.
 *
 * The user input and errors are cached.
 */
export interface IRegistrationFormCache {
  errors?: Partial<IRegistrationFormInput> & { responseError?: string };
  userInput?: IRegistrationFormInput;
}

interface IRegistrationFormProps {
  cache?: IRegistrationFormCache;
}

const validateRegistrationField = (
  registrationInput: IRegistrationFormInput
) => {
  // An object to store errors for all fields.
  const errors: Partial<IRegistrationFormInput> = {};

  // Check if the submitted given name is not empty.
  if (!registrationInput.givenName) {
    errors.givenName = "Given name can't be empty";
  }

  // Check if the submitted family name is not empty.
  if (!registrationInput.familyName) {
    errors.familyName = "Family name can't be empty";
  }

  // Check if the submitted username is not empty.
  if (!registrationInput.username) {
    errors.username = `Username cannot be empty.`;
  }

  // Check if the submitted username contains upper case characters.
  if (registrationInput.username !== registrationInput.username.toLowerCase()) {
    errors.username = "Username must be in lowercase";
  }

  // Check if the submitted email address is valid.
  if (!validate(registrationInput.email, { minDomainAtoms: 2 })) {
    errors.email = "Email address is invalid";
  }

  // Check if the password is secure enough.
  if (registrationInput.password.length <= 6) {
    errors.password = "Password should be at least 7 characters long";
  }

  // Check if passwords match
  if (registrationInput.password !== registrationInput.confirmPassword) {
    errors.confirmPassword = "Passwords do not match";
  }

  // return an array consisting of error message if any or empty.
  return errors;
};

const Form = styled.form.attrs({
  className: "column is-4 is-offset-4"
})`
  && {
    padding: 1px;
  }
`;

const Heading = styled.h1.attrs({
  className: "title is-3"
})`
  && {
    text-align: center;
    margin: 20px 0 20px 0;
  }
`;

const ErrorMessage = styled.p.attrs({
  className: "error tile"
})`
  && {
    text-align: left;
    margin: 3px 0 3px 0;
    padding: 2px 0 2px 0;
    background-color: hsl(348, 100%, 61%);
    color: white;
  }
`;

class RegistrationForm extends React.Component<
  IRegistrationFormProps,
  IRegistrationFormState
> {
  public state: IRegistrationFormState = {
    errors: {
      affiliation: "",
      confirmPassword: "",
      email: "",
      familyName: "",
      givenName: "",
      password: "",
      responseError: "",
      username: ""
    },
    registered: false,
    userInput: {
      affiliation: "",
      confirmPassword: "",
      email: "",
      familyName: "",
      givenName: "",
      password: "",
      username: ""
    }
  };

  /**
   * Populate the state from cached values.
   */
  componentDidMount() {
    if (this.props.cache) {
      this.setState(() => (this.props.cache as any) || {});
    }
  }

  onHandleSubmit = async (e: React.FormEvent<EventTarget>, signup: any) => {
    e.preventDefault();

    // Validate the user input fields
    const errors: object = validateRegistrationField(this.state.userInput);
    this.updateState({ errors });

    // Check if there is an error, if there is abort signing up.
    if (Object.keys(errors).length > 0) {
      return;
    }

    try {
      // If all fields are validated, sign up
      const userDetails = { ...this.state.userInput };
      delete userDetails.confirmPassword;
      await signup({ variables: userDetails });

      // Reset the state when registering succeeded
      this.updateState({
        registered: true,
        userInput: {
          affiliation: "",
          confirmPassword: "",
          email: "",
          familyName: "",
          givenName: "",
          password: "",
          username: ""
        }
      });

      alert("Successfully registered, Login using your username and password");
    } catch (error) {
      this.updateState({
        errors: {
          ...this.state.errors,
          responseError: error.message
            .replace("Network error: ", "")
            .replace("GraphQL error: ", "")
        }
      });
    }
  };

  onInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const name = e.target.name;
    const value = e.target.value;
    // Update form content with the new user input
    this.updateState({
      userInput: {
        ...this.state.userInput,
        [name]: value
      }
    });
  };

  render() {
    const { errors, registered } = this.state;
    const {
      affiliation,
      confirmPassword,
      email,
      familyName,
      givenName,
      password,
      username
    } = this.state.userInput;

    if (registered) {
      return <Redirect to={"/login"} />;
    }

    return (
      <Mutation mutation={SIGNUP_MUTATION}>
        {(signup: any, { loading }: any) => {
          return (
            <Form onSubmit={e => this.onHandleSubmit(e, signup)}>
              <fieldset disabled={loading} aria-disabled={loading}>
                <Heading>Create your account</Heading>
                {errors.responseError ? (
                  <ErrorMessage>{errors.responseError}</ErrorMessage>
                ) : null}

                {/* Given name */}
                <div className="field">
                  <label className="label">
                    Given name (first name)
                    <InputField
                      error={errors.givenName}
                      name={"givenName"}
                      onChange={this.onInputChange}
                      value={givenName}
                    />
                  </label>
                </div>

                {/* Family name */}
                <div className="field">
                  <label className="label">
                    Family name (surname)
                    <InputField
                      error={errors.familyName}
                      name={"familyName"}
                      onChange={this.onInputChange}
                      value={familyName}
                    />
                  </label>
                </div>

                {/* Email address */}
                <div className="field">
                  <label className="label">
                    Email address
                    <InputField
                      error={errors.email}
                      name={"email"}
                      onChange={this.onInputChange}
                      value={email}
                    />
                  </label>
                </div>

                {/* Username */}
                <div className="field">
                  <label className="label">
                    Username
                    <InputField
                      error={errors.username}
                      name={"username"}
                      onChange={this.onInputChange}
                      value={username}
                    />
                  </label>
                </div>

                {/* Affiliation */}
                <div className="field">
                  <label className="label">
                    Affiliation
                    <InputField
                      error={errors.affiliation}
                      name={"affiliation"}
                      placeholder={"E.g. University of Cape Town"}
                      onChange={this.onInputChange}
                      value={affiliation}
                    />
                  </label>
                </div>

                {/* Password */}
                <div className="field">
                  <label className="label">
                    Password
                    <InputField
                      error={errors.password}
                      name={"password"}
                      placeholder={"At least 7 characters"}
                      onChange={this.onInputChange}
                      type={"password"}
                      value={password}
                    />
                  </label>
                </div>

                {/* Retyped password */}
                <div className="field">
                  <label className="label">
                    Re-enter password
                    <InputField
                      error={errors.confirmPassword}
                      name={"confirmPassword"}
                      onChange={this.onInputChange}
                      type={"password"}
                      value={confirmPassword}
                    />
                  </label>
                </div>

                {/* Submit */}
                <button
                  className="button is-success is-fullwidth is-rounded"
                  data-test="signUp"
                  type={"submit"}
                >
                  {loading ? "Signing up..." : "Sign up"}
                </button>
              </fieldset>
            </Form>
          );
        }}
      </Mutation>
    );
  }

  /**
   * Update the form state and the cache.
   */
  private updateState = (update: object) => {
    this.setState(
      () => update,
      () => {
        if (this.props.cache) {
          this.props.cache.errors = _.cloneDeep(this.state.errors);
          this.props.cache.userInput = _.cloneDeep(this.state.userInput);
        }
      }
    );
  };
}

export default RegistrationForm;
