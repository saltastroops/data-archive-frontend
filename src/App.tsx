import * as React from "react";
import { Query } from "react-apollo";
import { Redirect, Route, Switch } from "react-router-dom";
import DataRequestsForm from "./components/dataRequest/DataRequestsForm";
import LoginForm, { ILoginFormCache } from "./components/LoginForm";
import NavigationBar from "./components/NavigationBar";
import RegistrationForm, {
  IRegistrationFormCache
} from "./components/RegistrationForm";
import ISearchFormCache from "./components/searchFormComponents/ISearchFormCache";
import SearchPage from "./components/searchFormComponents/SearchPage";
import UserUpdateForm, {
  IUserUpdateFormCache
} from "./components/UserUpdateForm";
import { USER_QUERY } from "./graphql/Query";

interface IUser {
  familyName: string;
  givenName: string;
  isAdmin: boolean;
}

interface IProtectedRouteProps {
  component: any;
  user: IUser | null | undefined;
  [propName: string]: any;
}

interface ICache {
  loginForm: ILoginFormCache;
  registrationForm: IRegistrationFormCache;
  searchForm: ISearchFormCache;
  userUpdateForm: IUserUpdateFormCache;
}

/**
 * A route which redirects to the login page if the user is not logged in.
 */
function ProtectedRoute({
  component: Component,
  user,
  ...rest
}: IProtectedRouteProps) {
  return (
    <Route
      {...rest}
      render={props =>
        user ? (
          <Component {...props} />
        ) : (
          <Redirect
            to={{
              pathname: "/login",
              state: { from: props.location }
            }}
          />
        )
      }
    />
  );
}

interface IAppState {
  user?: IUser;
  screenDimensions: { innerHeight: number; innerWidth: number };
}

/**
 * The data archive.
 */
class App extends React.Component<{}, IAppState> {
  state = {
    screenDimensions: {
      innerHeight: window.innerHeight,
      innerWidth: window.innerWidth
    },
    user: undefined
  };

  private cache: ICache = {
    loginForm: {},
    registrationForm: {},
    searchForm: {},
    userUpdateForm: {}
  };

  public render() {
    return (
      <Query query={USER_QUERY}>
        {({ data, loading }: any) => {
          if (loading) {
            return <p>Loading...</p>;
          }

          const currentUser =
            data && data.user
              ? {
                  familyName: data.user.familyName,
                  givenName: data.user.givenName,
                  isAdmin: data.user.roles.some(
                    (role: string) => role === "ADMIN"
                  )
                }
              : null;

          return (
            <>
              <NavigationBar user={currentUser} />

              <Switch>
                {/* search page */}
                <Route
                  exact={true}
                  path="/"
                  render={() => (
                    <SearchPage
                      cache={this.cache.searchForm}
                      screenDimensions={this.state.screenDimensions}
                    />
                  )}
                />

                {/* registration page */}
                <Route
                  exact={true}
                  path="/register"
                  render={() => (
                    <RegistrationForm cache={this.cache.registrationForm} />
                  )}
                />

                {/* login page */}
                <Route
                  exact={true}
                  path="/login"
                  render={() => <LoginForm cache={this.cache.loginForm} />}
                />

                {/* account details page */}
                <ProtectedRoute
                  user={currentUser}
                  exact={true}
                  path="/account"
                  component={() => (
                    <h1 className="title">
                      {currentUser
                        ? `${currentUser.givenName} ${
                            currentUser.familyName
                          } Account`
                        : "No User"}
                    </h1>
                  )}
                />

                {/* data requests page */}
                <ProtectedRoute
                  user={currentUser}
                  exact={true}
                  path="/data-requests"
                  component={() => <DataRequestsForm />}
                />

                {/* cart page */}
                <Route
                  exact={true}
                  path="/cart"
                  component={() => <h1 className="title">Cart page</h1>}
                />

                {/* admin page */}
                <ProtectedRoute
                  user={currentUser}
                  exact={true}
                  path="/admin"
                  component={() => <h1 className="title">Admin page</h1>}
                />

                {/* update user page */}
                <ProtectedRoute
                  user={currentUser}
                  exact={true}
                  path="/user-update"
                  component={() => (
                    <UserUpdateForm cache={this.cache.userUpdateForm} />
                  )}
                />

                {/* page not found */}
                <Route
                  component={() => <h1 className="title">Page not found</h1>}
                />
              </Switch>
            </>
          );
        }}
      </Query>
    );
  }
}

export default App;
