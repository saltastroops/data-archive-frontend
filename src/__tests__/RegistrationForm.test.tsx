import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import { MockedProvider } from "react-apollo/test-utils";
import { MemoryRouter } from "react-router";
import wait from "waait";
import App from "../App";
import RegistrationForm from "../components/RegistrationForm";
import { SIGNUP_MUTATION } from "../graphql/Mutations";
import click from "../util/click";

window.matchMedia = jest.fn().mockImplementation(query => {
  return {
    addListener: jest.fn(),
    matches: false,
    media: query,
    onchange: null,
    removeListener: jest.fn()
  };
});

// Helper function for simulating input field value change.
function inputTyping(wrapper: any, name: string, value: string) {
  wrapper.find(`input[name="${name}"]`).simulate("change", {
    target: { name, value }
  });
}

// Updated userInput state
const updatedState = {
  errors: {
    affiliation: "",
    confirmPassword: "",
    email: "",
    familyName: "",
    givenName: "",
    password: "",
    username: ""
  },
  userInput: {
    affiliation: "University of Cape Town",
    confirmPassword: "securepassword",
    email: "valid@email.address",
    familyName: "Smith",
    givenName: "John",
    password: "securepassword",
    username: "sj"
  }
};

describe("RegistrationForm Component", () => {
  it("renders the RegistrationForm having unpopulated props with no errors", () => {
    // RegistrationForm component wrapper.
    const wrapper = mount(
      <MockedProvider>
        <RegistrationForm />
      </MockedProvider>
    );

    // Expect the snapshot to match the RegistrationForm component.
    expect(toJson(wrapper.find("form"))).toMatchSnapshot();
  });

  it("successfully submits valid input", async () => {
    const register = jest.fn();
    const userDetails = {
      affiliation: "University of Cape Town",
      email: "valid@email.address",
      familyName: "Smith",
      givenName: "John",
      password: "securepassword",
      username: "sj"
    };
    const mocks = [
      {
        request: {
          query: SIGNUP_MUTATION,
          variables: userDetails
        },
        result: () => {
          register();
          return {
            data: {
              signup: {
                __typename: "User",
                ...userDetails,
                id: "1"
              }
            }
          };
        }
      }
    ];

    // RegistrationForm component wrapper.
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <MemoryRouter initialEntries={["/register"]}>
          <RegistrationForm />
        </MemoryRouter>
      </MockedProvider>
    );

    // RegistrationForm component instance
    const instance = wrapper.find("RegistrationForm").instance() as any;

    // Spy on the setState function.
    const setState = jest.spyOn(instance, "setState");

    // Simulate state change when the affiliation input field value changes.
    inputTyping(wrapper, "affiliation", userDetails.affiliation);
    // Simulate state change when the confirmPassword input field value changes.
    inputTyping(wrapper, "confirmPassword", userDetails.password);
    // Simulate state change when the emailAddress input field value changes.
    inputTyping(wrapper, "email", userDetails.email);
    // Simulate state change when the familyName input field value changes.
    inputTyping(wrapper, "familyName", userDetails.familyName);
    // Simulate state change when the givenName input field value changes.
    inputTyping(wrapper, "givenName", userDetails.givenName);
    // Simulate state change when the password input field value changes.
    inputTyping(wrapper, "password", userDetails.password);
    // Simulate state change when the username input field value changes.
    inputTyping(wrapper, "username", userDetails.username);

    // Expect setState to have been called
    expect(setState.mock.calls.length).toBe(7);

    // Expect the property affiliation of the state to have been updated with the correct value.
    expect(instance.state.userInput.affiliation).toBe(
      "University of Cape Town"
    );

    // Expect the property confirmPassword of the state to have been updated with the correct value.
    expect(instance.state.userInput.confirmPassword).toBe("securepassword");

    // Expect the property emailAddress of the state to have been updated with the correct value.
    expect(instance.state.userInput.email).toBe("valid@email.address");

    // Expect the property familyName of the state to have been updated with the correct value.
    expect(instance.state.userInput.familyName).toBe("Smith");

    // Expect the property givenName of the state to have been updated with the correct value.
    expect(instance.state.userInput.givenName).toBe("John");

    // Expect the property password of the state to have been updated with the correct value.
    expect(instance.state.userInput.password).toBe("securepassword");

    // Expect the property username of the state to have been updated with the correct value.
    expect(instance.state.userInput.username).toBe("sj");

    // Expect the button not to indicate loading
    expect(wrapper.find('[data-test="signUp"]').text()).toContain("Sign up");

    // The mutation has not been called (yet)
    expect(register).not.toHaveBeenCalled();

    // Simulate the submitting of the form.
    wrapper.find('[data-test="signUp"]').simulate("submit");

    // Expect the button to indicate loading
    expect(wrapper.find('[data-test="signUp"]').text()).toContain("Signing up");

    // The mutation has been called
    await wait(50); // 50 rather than 0 to give a potential state update the chance to finish
    wrapper.update();
    expect(register).toHaveBeenCalled();

    // Expect no error message
    expect(wrapper.find("p").length).toBe(0);
  });

  it("displays an error message if an invalid given name is submitted", () => {
    // RegistrationForm component wrapper.
    const wrapper = mount(
      <MockedProvider>
        <RegistrationForm />
      </MockedProvider>
    );

    // RegistrationForm component instance
    const instance = wrapper.find("RegistrationForm").instance() as any;

    // Set the state with valid information
    instance.setState({ userInput: updatedState.userInput });

    // Spy on the setState function.
    const setState = jest.spyOn(instance, "setState");

    // Simulate state change when the givenName input field value changes.
    inputTyping(wrapper, "givenName", "");

    // Expect setState to have been called
    expect(setState.mock.calls.length).toBe(1);

    // Simulate the submiting of the form.
    wrapper.find('[data-test="signUp"]').simulate("submit");

    // Expect an error message.
    expect(wrapper.find("p").length).toBe(1);

    // Expect meaningful error message
    expect(wrapper.find("p").text()).toContain("Given name");
    expect(wrapper.find("p").text()).toContain("empty");
  });

  it("displays an error message if an invalid family name is submitted", () => {
    // RegistrationForm component wrapper.
    const wrapper = mount(
      <MockedProvider>
        <RegistrationForm />
      </MockedProvider>
    );

    // RegistrationForm component instance
    const instance = wrapper.find("RegistrationForm").instance() as any;

    // Set the state with valid information
    instance.setState({ userInput: updatedState.userInput });

    // Spy on the setState function.
    const setState = jest.spyOn(instance, "setState");

    // Simulate state change when the familyName input field value changes.
    inputTyping(wrapper, "familyName", "");

    // Expect setState to have been called
    expect(setState.mock.calls.length).toBe(1);

    // Simulate the submiting of the form.
    wrapper.find('[data-test="signUp"]').simulate("submit");

    // Expect an error message.
    expect(wrapper.find("p").length).toBe(1);

    // Expect meaningful error message
    expect(wrapper.find("p").text()).toContain("Family name");
    expect(wrapper.find("p").text()).toContain("empty");
  });

  it("displays an error message if an invalid family name is submitted", () => {
    // RegistrationForm component wrapper.
    const wrapper = mount(
      <MockedProvider>
        <RegistrationForm />
      </MockedProvider>
    );

    // RegistrationForm component instance
    const instance = wrapper.find("RegistrationForm").instance() as any;

    // Set the state with valid information
    instance.setState({ userInput: updatedState.userInput });

    // Spy on the setState function.
    const setState = jest.spyOn(instance, "setState");

    // Simulate state change when the emailAddress input field value changes.
    inputTyping(wrapper, "email", "invalid$email&address");

    // Expect setState to have been called
    expect(setState.mock.calls.length).toBe(1);

    // Simulate the submiting of the form.
    wrapper.find('[data-test="signUp"]').simulate("submit");

    // Expect an error message.
    expect(wrapper.find("p").length).toBe(1);

    // Expect meaningful error message
    expect(wrapper.find("p").text()).toContain("Email address");
    expect(wrapper.find("p").text()).toContain("invalid");
  });

  it("displays an error message if an invalid username is submitted", () => {
    // RegistrationForm component wrapper.
    const wrapper = mount(
      <MockedProvider>
        <RegistrationForm />
      </MockedProvider>
    );

    // RegistrationForm component instance
    const instance = wrapper.find("RegistrationForm").instance() as any;

    // Set the state with valid information
    instance.setState({ userInput: updatedState.userInput });

    // Spy on the setState function.
    const setState = jest.spyOn(instance, "setState");

    // Simulate state change when the username input field value changes.
    inputTyping(wrapper, "username", "sJ");

    // Expect setState to have been called
    expect(setState.mock.calls.length).toBe(1);

    // Simulate the submiting of the form.
    wrapper.find('[data-test="signUp"]').simulate("submit");

    // Expect an error message.
    expect(wrapper.find("p").length).toBe(1);

    // Expect meaningful error message
    expect(wrapper.find("p").text()).toContain("Username");
    expect(wrapper.find("p").text()).toContain("lowercase");
  });

  it("displays an error message if an invalid password is submitted", () => {
    // RegistrationForm component wrapper.
    const wrapper = mount(
      <MockedProvider>
        <RegistrationForm />
      </MockedProvider>
    );

    // RegistrationForm component instance
    const instance = wrapper.find("RegistrationForm").instance() as any;

    // Set the state with valid information
    instance.setState({ userInput: updatedState.userInput });

    // Spy on the setState function.
    const setState = jest.spyOn(instance, "setState");

    // Simulate state change when the password input field value changes.
    inputTyping(wrapper, "password", "short");

    // Expect setState to have been called
    expect(setState.mock.calls.length).toBe(1);

    // Simulate the submiting of the form.
    wrapper.find('[data-test="signUp"]').simulate("submit");

    // Expect an error message.
    expect(wrapper.find("p").length).toBe(2);

    // Expect meaningful error message
    expect(
      wrapper
        .find("p")
        .first()
        .text()
    ).toContain("Password");
    expect(
      wrapper
        .find("p")
        .first()
        .text()
    ).toContain("7 characters");
  });

  it("displays an error message if an invalid confirmed password is submitted", () => {
    // RegistrationForm component wrapper.
    const wrapper = mount(
      <MockedProvider>
        <RegistrationForm />
      </MockedProvider>
    );

    // RegistrationForm component instance
    const instance = wrapper.find("RegistrationForm").instance() as any;

    // Set the state with valid information
    instance.setState({ userInput: updatedState.userInput });

    // Spy on the setState function.
    const setState = jest.spyOn(instance, "setState");

    // Simulate state change when the password input field value changes.
    inputTyping(wrapper, "password", "securepassword");
    // Simulate state change when the confirmPassword input field value changes.
    inputTyping(wrapper, "confirmPassword", "notmatching");

    // Expect setState to have been called
    expect(setState.mock.calls.length).toBe(2);

    // Simulate the submiting of the form.
    wrapper.find('[data-test="signUp"]').simulate("submit");

    // Expect an error message.
    expect(wrapper.find("p").length).toBe(1);

    // Expect meaningful error message
    expect(
      wrapper
        .find("p")
        .first()
        .text()
    ).toContain("Password");
    expect(
      wrapper
        .find("p")
        .first()
        .text()
    ).toContain("do not");
  });

  it("should display an error if the registration fails", async () => {
    const register = jest.fn();
    const userDetails = {
      affiliation: "University of Cape Town",
      email: "valid@email.address",
      familyName: "Smith",
      givenName: "John",
      password: "securepassword",
      username: "sj"
    };
    const mocks = [
      {
        error: new Error("The server is having a coffee break!"),
        request: {
          query: SIGNUP_MUTATION,
          variables: userDetails
        }
      }
    ];

    // RegistrationForm component wrapper.
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <RegistrationForm />
      </MockedProvider>
    );

    // Simulate state change when the affiliation input field value changes.
    inputTyping(wrapper, "affiliation", userDetails.affiliation);
    // Simulate state change when the confirmPassword input field value changes.
    inputTyping(wrapper, "confirmPassword", userDetails.password);
    // Simulate state change when the emailAddress input field value changes.
    inputTyping(wrapper, "email", userDetails.email);
    // Simulate state change when the familyName input field value changes.
    inputTyping(wrapper, "familyName", userDetails.familyName);
    // Simulate state change when the givenName input field value changes.
    inputTyping(wrapper, "givenName", userDetails.givenName);
    // Simulate state change when the password input field value changes.
    inputTyping(wrapper, "password", userDetails.password);
    // Simulate state change when the username input field value changes.
    inputTyping(wrapper, "username", userDetails.username);

    // There is no error (yet).
    expect(wrapper.find("p.error").length).toBe(0);

    // Simulate the submiting of the form.
    wrapper.find('[data-test="signUp"]').simulate("submit");

    // There is an error now
    await wait(50); // 50 ms rather than 0 ms to give the state update a chance to finish
    wrapper.update();
    expect(wrapper.find("p.error").length).toBe(1);
    expect(wrapper.find("p.error").text()).toEqual(
      "The server is having a coffee break!"
    );
  });

  it("should cache values and errors", async () => {
    const wrapper = mount(
      <MockedProvider>
        <MemoryRouter>
          <App />
        </MemoryRouter>
      </MockedProvider>
    );

    await wait(0);
    wrapper.update();

    // Navigate to the registration form
    const registrationFormLink = wrapper.find('a[href="/register"]').first();
    click(registrationFormLink);

    await wait(0);
    wrapper.update();

    // Fill in an invalid email
    inputTyping(wrapper, "email", "invalid email");

    // Submit the form
    const submitButton = wrapper.find('[data-test="signUp"]');
    submitButton.simulate("submit");

    await wait(0);
    wrapper.update();

    // The value has been stored in the state, and there is an error
    const registrationFormState: any = wrapper.find("RegistrationForm").state();
    const emailValue = registrationFormState.userInput.email;
    expect(emailValue).toEqual("invalid email");
    const emailErrorMessage = registrationFormState.errors.email;
    expect(emailErrorMessage.length).toBeGreaterThan(0);

    // Navigate away from the registration form
    const cartLink = wrapper.find('a[href="/cart"]').first();
    click(cartLink);

    await wait(0);
    wrapper.update();

    // No login form any longer
    expect(wrapper.find("RegistrationForm").length).toBe(0);

    // Navigate back to the search form
    click(registrationFormLink);

    await wait(0);
    wrapper.update();

    // The values and errors have been re-inserted.
    const newRegistrationFormState: any = wrapper
      .find("RegistrationForm")
      .state();
    expect(newRegistrationFormState.userInput.email).toEqual(emailValue);
    expect(newRegistrationFormState.errors.email).toEqual(emailErrorMessage);
  });
});
