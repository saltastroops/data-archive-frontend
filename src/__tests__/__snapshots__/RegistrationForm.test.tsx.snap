// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationForm Component renders the RegistrationForm having unpopulated props with no errors 1`] = `
.c0.c0 {
  padding: 1px;
}

.c1.c1 {
  text-align: center;
  margin: 20px 0 20px 0;
}

<form
  className="column is-4 is-offset-4 c0"
  onSubmit={[Function]}
>
  <fieldset
    aria-disabled={false}
    disabled={false}
  >
    <styled.h1>
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [
              Object {
                "className": "title is-3",
              },
            ],
            "componentStyle": ComponentStyle {
              "componentId": "sc-dnqmqq",
              "isStatic": true,
              "lastClassName": "c1",
              "rules": Array [
                "
  && {
    text-align: center;
    margin: 20px 0 20px 0;
  }
",
              ],
            },
            "displayName": "styled.h1",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-dnqmqq",
            "target": "h1",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
      >
        <h1
          className="title is-3 c1"
        >
          Create your account
        </h1>
      </StyledComponent>
    </styled.h1>
    <div
      className="field"
    >
      <label
        className="label"
      >
        Given name (first name)
        <InputField
          error=""
          name="givenName"
          onChange={[Function]}
          value=""
        >
          <div
            className="control"
          >
            <input
              className="input "
              name="givenName"
              onChange={[Function]}
              value=""
            />
          </div>
        </InputField>
      </label>
    </div>
    <div
      className="field"
    >
      <label
        className="label"
      >
        Family name (surname)
        <InputField
          error=""
          name="familyName"
          onChange={[Function]}
          value=""
        >
          <div
            className="control"
          >
            <input
              className="input "
              name="familyName"
              onChange={[Function]}
              value=""
            />
          </div>
        </InputField>
      </label>
    </div>
    <div
      className="field"
    >
      <label
        className="label"
      >
        Email address
        <InputField
          error=""
          name="email"
          onChange={[Function]}
          value=""
        >
          <div
            className="control"
          >
            <input
              className="input "
              name="email"
              onChange={[Function]}
              value=""
            />
          </div>
        </InputField>
      </label>
    </div>
    <div
      className="field"
    >
      <label
        className="label"
      >
        Username
        <InputField
          error=""
          name="username"
          onChange={[Function]}
          value=""
        >
          <div
            className="control"
          >
            <input
              className="input "
              name="username"
              onChange={[Function]}
              value=""
            />
          </div>
        </InputField>
      </label>
    </div>
    <div
      className="field"
    >
      <label
        className="label"
      >
        Affiliation
        <InputField
          error=""
          name="affiliation"
          onChange={[Function]}
          placeholder="E.g. University of Cape Town"
          value=""
        >
          <div
            className="control"
          >
            <input
              className="input "
              name="affiliation"
              onChange={[Function]}
              placeholder="E.g. University of Cape Town"
              value=""
            />
          </div>
        </InputField>
      </label>
    </div>
    <div
      className="field"
    >
      <label
        className="label"
      >
        Password
        <InputField
          error=""
          name="password"
          onChange={[Function]}
          placeholder="At least 7 characters"
          type="password"
          value=""
        >
          <div
            className="control"
          >
            <input
              className="input "
              name="password"
              onChange={[Function]}
              placeholder="At least 7 characters"
              type="password"
              value=""
            />
          </div>
        </InputField>
      </label>
    </div>
    <div
      className="field"
    >
      <label
        className="label"
      >
        Re-enter password
        <InputField
          error=""
          name="confirmPassword"
          onChange={[Function]}
          type="password"
          value=""
        >
          <div
            className="control"
          >
            <input
              className="input "
              name="confirmPassword"
              onChange={[Function]}
              type="password"
              value=""
            />
          </div>
        </InputField>
      </label>
    </div>
    <button
      className="button is-success is-fullwidth is-rounded"
      data-test="signUp"
      type="submit"
    >
      Sign up
    </button>
  </fieldset>
</form>
`;
